### PR TITLE
Take `impl Into<DefId>` for query methods on `TyCtxt`

### PR DESCRIPTION
**Pull request by ecstatic-morse**
Alternative implementation of #70956. cc #70853.

### CODE REVIEW

**Comment by ecstatic-morse** [comment]
cc @marmeladema @eddyb

File: src/librustc_middle/ty/query/plumbing.rs [hunk 1]
```diff
@@ -234,18 +234,54 @@ macro_rules! hash_result {
...
 }
 
+macro_rules! define_query_getter {
+    (TyCtxtAt<$tcx:tt>, $(#[$attr:meta])* $name:ident(DefId) -> $V:ty) => {
+        $(#[$attr])*
+        #[inline(always)]
+        pub fn $name(self, key: impl Into<DefId>) -> $V {
+            fn inner(this: TyCtxtAt<$tcx>, key: DefId) -> $V {
+                get_query::<queries::$name<'_>, _>(this.tcx, this.span, key)
+            }
```
Comment (eddyb) [inline src/librustc_middle/ty/query/plumbing.rs:252]: Is `get_query` inlined? This does nothing otherwise.

File: src/librustc_middle/ty/query/plumbing.rs [hunk 2]
```diff
@@ -234,18 +234,54 @@ macro_rules! hash_result {
...
+            $($( $(#[$attr])* category<$category> [$($modifiers)*] fn $name: $node($($K)*) -> $V,)*)*
         }
     }
 }
 
+macro_rules! define_query_getter {
+    (TyCtxtAt<$tcx:tt>, $(#[$attr:meta])* $name:ident(DefId) -> $V:ty) => {
+        $(#[$attr])*
+        #[inline(always)]
+        pub fn $name(self, key: impl Into<DefId>) -> $V {
```
Comment (eddyb) [inline src/librustc_middle/ty/query/plumbing.rs:249]: Can you make a type macro and use *that* to pattern-match on just the key type?
I think we can just use `.into()` in all methods, as it will compile to a noop when the types are the same.

File: src/librustc_middle/ty/query/plumbing.rs (hunk 1 above)
Comment (ecstatic-morse) [reply src/librustc_middle/ty/query/plumbing.rs]: This was an attempt to address [@Zoxc's comment](https://github.com/rust-lang/rust/pull/70956#discussion_r406293448). I don't understand why it's beneficial for the caller of `get_query` to be monomorphic, since it is `#[inline(never)]`. Perhaps I misunderstood?

File: src/librustc_middle/ty/query/plumbing.rs (hunk 2 above)
Comment (ecstatic-morse) [reply src/librustc_middle/ty/query/plumbing.rs:249]: That's a good idea. It won't work if we do want a monomorphic helper function, but maybe we don't.

**Comment by nikomatsakis** [comment]
Per rust-lang/team#316, r? @eddyb -- let me know @eddyb if you'd prefer someone else.

File: src/librustc_middle/ty/query/plumbing.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_middle/ty/query/plumbing.rs:249]: `get_query` is `#[inline(never)]` (I just checked) so you don't need this wrapper.

**Comment by ecstatic-morse** [comment]
@eddyb The last two commits implement your suggestion for simplifying the macro and replace `Into` with a newly defined `IntoQueryParam` respectively. You can't seal traits with blanket impls, however.

File: src/librustc_middle/ty/query/mod.rs [hunk 3]
```diff
@@ -189,3 +189,23 @@ pub fn force_from_dep_node<'tcx>(tcx: TyCtxt<'tcx>, dep_node: &DepNode) -> bool
...
     rustc_dep_node_try_load_from_on_disk_cache!(dep_node, tcx)
 }
+
+/// An analogue of the `Into` trait that's intended only for query paramaters.
+///
+/// This exists to allow queries to accept either `DefId` or `LocalDefId` while requiring that the
+/// user call `to_def_id` to convert between them everywhere else.
+pub trait IntoQueryParam<P> {
+    fn into_query_param(self) -> P;
+}
```
Comment (eddyb) [inline src/librustc_middle/ty/query/mod.rs:199]: In case there was any confusion, "sealing" works like this:
```rust
mod sealed {
    pub trait IntoQueryParam<P> {
        fn into_query_param(self) -> P;
    }
}
```

Then you can refer to it as `sealed::IntoQueryParam`, and as long as `sealed` remains private, and you don't reexport `IntoQueryParam` somewhere, it should be impossible to implement or import the trait, outside of the parent module of the `mod sealed`.

File: src/librustc_span/def_id.rs [hunk 4]
```diff
@@ -232,6 +232,12 @@ impl fmt::Debug for LocalDefId {
     }
 }
 
+impl From<LocalDefId> for DefId {
+    fn from(v: LocalDefId) -> Self {
+        v.to_def_id()
+    }
+}
```
Comment (eddyb) [inline src/librustc_span/def_id.rs:239]: This can be removed now, right?

File: src/librustc_middle/ty/query/mod.rs (hunk 3 above)
Comment (ecstatic-morse) [reply src/librustc_middle/ty/query/mod.rs:199]: Ah, I thought you meant the super-trait patten that's in the API guidelines.

Comment (eddyb) [reply src/librustc_middle/ty/query/mod.rs]: @ecstatic-morse ah I see, that pattern uses a sealed trait for a specific purpose.
Or maybe I'm the only person referring to "`pub` but in a private `mod`" as "sealed".

File: src/librustc_middle/ty/query/plumbing.rs (hunk 1 above)
Comment (cjgillot) [reply src/librustc_middle/ty/query/plumbing.rs]: IFAIK, the LLVM codegen of the `{get,force,ensure}_query` functions is responsible for most of the compile time for librustc_middle. It happens once for each of the 170 queries. By making the caller generic, part of this codegen may be replicated in caller crates. This would have a serious impact on rustc's compile time. Please correct me if `#[inline(never)]` is enough to prevent this.

**Review by eddyb — APPROVED** [review]

**Comment by eddyb** [comment]
LGTM, but before I approve:

@bors try @rust-timer queue

File: src/librustc_middle/ty/query/plumbing.rs (hunk 1 above)
Comment (eddyb) [reply src/librustc_middle/ty/query/plumbing.rs]: The wrapper functions are already `#[inline(always)]` so they would be codegen'd cross-crate even when not generic.

File: src/librustc_middle/ty/query/plumbing.rs [hunk 5]
```diff
@@ -380,8 +385,8 @@ macro_rules! define_queries_inner {
         impl TyCtxtEnsure<$tcx> {
             $($(#[$attr])*
             #[inline(always)]
-            pub fn $name(self, key: $K) {
-                ensure_query::<queries::$name<'_>, _>(self.tcx, key)
+            pub fn $name(self, key: query_helper_param_ty!($($K)*)) {
+                ensure_query::<queries::$name<'_>, _>(self.tcx, key.into_query_param())
```
Comment (eddyb) [inline src/librustc_middle/ty/query/plumbing.rs:389]: Actually, are these changes needed? "ensuring" should be rather rare AFAIK?

**Comment by marmeladema** [comment]
@ecstatic-morse @eddyb is the performance good enough for this to land?

**Comment by ecstatic-morse** [comment]
@marmeladema Too slow IMO. We'll need to micro-optimize or abandon this idea.

File: src/librustc_middle/ty/query/mod.rs [hunk 6]
```diff
@@ -189,3 +189,29 @@ pub fn force_from_dep_node<'tcx>(tcx: TyCtxt<'tcx>, dep_node: &DepNode) -> bool
...
+    }
+
+    impl<P> IntoQueryParam<P> for P {
+        fn into_query_param(self) -> P {
+            self
+        }
+    }
+
+    impl IntoQueryParam<DefId> for LocalDefId {
+        fn into_query_param(self) -> DefId {
```
Comment (eddyb) [inline src/librustc_middle/ty/query/mod.rs:211]: Try adding `#[inline(always)]` to this (and `to_def_id` if it doesn't have it).

Although, wait, are we using this at all now? Is the slowdown *just* from going through the noop impl?

**Comment by ecstatic-morse** [comment]
@bors try
@rust-timer queue

File: src/librustc_middle/ty/query/mod.rs (hunk 6 above)
Comment (ecstatic-morse) [reply src/librustc_middle/ty/query/mod.rs:213]: It's used exactly once as a test 6e5a210.

**Comment by eddyb** [comment]
Heh that's what it was! r=me if there is no disagreement about doing this.

**Comment by eddyb** [comment]
Eh let's just do it.

@bors r+
